packages/runtime/container-runtime/src/channelCollection.ts

```diff
@@ -579,8 +579,8 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 		 * If the container is detached, this data store will be part of the summary that makes the container attached.
 		 */
 		if (this.parentContext.attachState !== AttachState.Detached) {
-			localContext.setAttachState(AttachState.Attaching);
 			this.submitAttachChannelOp(localContext);
+			localContext.setAttachState(AttachState.Attaching);
```

Comment:
this is the primary change, and i think it is safe, well at least as safe as it was before. this also matches how the container works, specifically, get capture the summary, and then transition to attaching. Since this code is all synchronous the risk of something bad happening is small, or about the same as before.

Regardless of the order having the summary retrieval decoupled from the state transition leaves open the possibility for bugs. With this change it is possible the dds state is modified after the summary is captured, so it won't be in the summary or in ops, only in memory. Conversely, the in the existing code if a change happened after the attach state change it would get an op, but it would also be in the summary, so it gets duplicated. Both of these are data corruption. 

The better and much bigger fix would be to get the summary as part of the transition, so as each ddses attachment summary is retrieved it then transition to attaching. this would be a substantial refactoring, and due to the code being synchronous the opportunity for these types of issues is small, and would likely represent a framework issue.

Reply:
Nice. This is exactly the kind of change I was wondering about when we were discussing this last week.

@agarwal-navin - thoughts? In the existing code, handles serialized (and bound) while generating the attach summary can result in redundant attach ops.

Reply:
Looks like a nice and safe change.
I had update the attach / visibility state some time back and added some asserts that should catch if anything wrong was happening. However, I just realized that I had left some workaround in place due to back compat which were never removed. We should remove these 2 attachGraph() calls and then check if this change still passes all tests.
https://github.com/microsoft/FluidFramework/blob/64662e10aeb68c68b9ff2aafb9b393c45c27d34b/packages/runtime/datastore/src/dataStoreRuntime.ts#L1221
https://github.com/microsoft/FluidFramework/blob/64662e10aeb68c68b9ff2aafb9b393c45c27d34b/packages/runtime/datastore/src/dataStoreRuntime.ts#L936

Reply:
i ran into problem when i tried to remove those, so let do that as a follow up